[feat] Gossip/SlowMo (#378)
facebookresearch/fairscale: docs/source/deep_dive/slowmo_ddp.rst

1. Compute the forward pass.

2. Compute the backward pass.

3. During the backward pass, using a backward hook, the gradients are synchronized using allreduce across the different GPUs on a node.

4. Perform the ``optimizer.step()`` to update parameters on a node with the gradients of that node.

5. Approximately average the parameters using a base algorithm - one of LocalSGD or SGP (both are described above).

6. Perform the slow momentum update step once every ``slowmo_frequency`` (typically 48) iterations. In this step, the parameters on different
   nodes are reduced, followed by a ``slowmo_optimizer.step()``. Note that this ``slowmo_optimizer`` is different from the original optimizer,
   and it is done in a `Zero-1 <https://fairscale.readthedocs.io/en/latest/deep_dive/oss_sdp_fsdp.html>`_ like manner to save memory.

Best practices for using ``SlowMoDistributedDataParallel``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1. SlowMo will be useful in deep learning workloads which run on greater than 2 nodes in clusters with a slow interconnect, eg Ethernet.

2. SlowMo should be useful in your workload if the following condition holds:

>    :math:`\textrm{time_taken_for_all_reduce_of_gradients} \times (1 - \frac{1}{\textrm{localsgd_frequency}} ) > \textrm{time_taken_for_backward_pass}`

Reviewer: Actually the formula was originally -
time_taken_for_all_reduce > time_taken_for_backward_pass + time_taken_for_all_reduce/ localsgd_frequency

I shortened it to make it easier to evaluate for someone evaluating the condition (which has a tradeoff of making it harder to understand)

time_taken_for_backward_pass is extra on the right hand side because we do not overlap the backward pass with communication on SlowMo

>    :math:`\textrm{time_taken_for_all_reduce_of_gradients} \times (1 - \frac{1}{\textrm{localsgd_frequency}} ) > \textrm{time_taken_for_backward_pass}`

Reviewer: I could add a short explanation as -

`
   The formula above is a simplified version of:
   :math:`\textrm{time_taken_for_all_reduce_of_gradients} > \textrm{time_taken_for_backward_pass} + \frac{\textrm{time_taken_for_all_reduce_of_gradients}}{\textrm{localsgd_frequency}} )`
   ``time_taken_for_backward_pass`` is extra on the right hand side because we do not overlap the backward pass with communication in the current implementation of SlowMo.
`

I'll format it a little better so that it doesn't appear too cluttered to the user

   In clusters with slower interconnect, ``time_taken_for_all_reduce_of_gradients`` will go up, leading to SlowMo being more useful. ``localsgd_frequency``
   is also an important factor here. More details on varying that to affect performance are in tip 2 of
   `Performance tips for SlowMoDistributedDataParallel`_.

> 3. ``slowmo_momentum`` will need to be tuned for obtaining good accuracy. A random search across 4 values from [0.1, 0.2, ..., 0.7] should be good enough

Reviewer 2: Should it be 7 values?

Reviewer: This is what we usually did earlier. I guess I could also change this to a random search for 4 values between 0.0 and 0.7

Reviewer 2: Oh are you saying you were randomly picking 4 values among `{0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7}` and keeping the best one? That doesn't make much sense IMO, because:
* Assuming this is uniform iid sampling, there's a good chance you'll have duplicates
* If you're unlucky you may miss some of the extremes

If you are going to try only 4 values out of 7, just do a grid search, not a random search. I'm not sure what kind of value you typically ended up with, but maybe a grid search other the (5) values `{0.1, 0.2, 0.3, 0.5, 0.7}` would be a reasonable recommendation?

   for tuning. This ``slowmo_momentum`` value holds consistent across multiple runs with similar settings.  When the number of nodes used is increased,
   however, a higher value of ``slow_momentum`` should be needed. More details about this can be found in the
   `documentation <https://fairscale.readthedocs.io/en/latest/api/experimental/nn/slowmo_ddp.html>`_.

> 4. Adding SlowMo involves two steps, which can be found in the `tutorial <https://fairscale.readthedocs.io/en/latest/tutorials/slowmo_ddp.html>`_.

Reviewer 2: ```suggestion
4. Adding SlowMo to existing Distributed Data Parallel code involves two steps, which can be found in the `tutorial <https://fairscale.readthedocs.io/en/latest/tutorials/slowmo_ddp.html>`_.
```

Also, can internal links be generated by the doc tool instead of being hardcoded? (this way they would remain valid for versioned docs and also if the website changes). This comment applies to all internal links.

Performance tips for ``SlowMoDistributedDataParallel``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> 1. ``nprocs_per_node`` should be set to the number of GPUs in a node. This allows the API to exploit the fast interconnect between different GPUs

Reviewer 2: ```suggestion
1. ``nprocs_per_node`` should be set to the number of GPUs on each node. This allows the API to exploit the fast interconnect between different GPUs
```

   on a node.

2. Increasing the ``localsgd_frequency`` results in an increase in speed. However, it comes with a tradeoff of reducing the accuracy.
   We recommend keeping the ``localsgd_frequency`` at 3.

> 3. ``slowmo_memory_efficient`` should typically be used. It reduces memory usage by sharding the extra slow momentum optimizer's parameters in

Reviewer 2: ```suggestion
3. ``slowmo_memory_efficient`` should typically be used (this is the default behavior). It reduces memory usage by sharding the additional slow momentum optimizer's parameters in
```
(replacing "extra" by "additional" because otherwise "extra slow" sounds like something very very slow)
